Refactor and inclusion of `te_rmsnorm_fwd_triton`

Micky774 opened the pull request:
# Description

This PR:
- Introduces `te_rmsnorm_fwd_triton` which is refactored to match `tex.rmsnorm_fwd` in API and functionality.
- Adds tests confirming API compatibility and functional exactness of `te_rmsnorm_fwd_triton` and `tex.rmsnorm_fwd`
- Includes the use of `te_rmsnorm_fwd_triton` in places where `IS_HIP_EXTENSION` and `NVTE_USE_RMSNORM_TRITON` are appropriately set.

Partially addresses https://github.com/ROCm/frameworks-internal/issues/12520

Note: I will open a follow-up PR to include FP8 support.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [x] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Introduces `te_rmsnorm_fwd_triton` which is refactored to match `tex.rmsnorm_fwd` in API and functionality.
- Adds tests confirming API compatibility and functional exactness of `te_rmsnorm_fwd_triton` and `tex.rmsnorm_fwd`
- Includes the use of `te_rmsnorm_fwd_triton` in places where `IS_HIP_EXTENSION` and `NVTE_USE_RMSNORM_TRITON` are appropriately set.

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

Micky774 committed:
Initial addition of refactored te_rms_forward function

Micky774 committed:
Formatting

wangye805 commented on tests/pytorch/triton_kernels/test_rmsnorm.py:
```diff
@@ -113,3 +114,57 @@ def test_rmsnorm_bwd_triton(M, N, in_dtype, out_dtype, zero_centered_gamma):
     )
 
 # TODO: bring back fwd pytests after refactoring te_rmsnorm_fwd_triton
```
This todo can be removed

wangye805 commented on transformer_engine/pytorch/module/_common.py:
```diff
@@ -25,14 +25,14 @@
     fwd_normalization_funcs = {
         "LayerNorm": tex.layernorm_fwd,
-        "RMSNorm": tex.rmsnorm_fwd,
+        "RMSNorm": te_rmsnorm_fwd_triton if use_rmsnorm_triton else tex.rmsnorm_fwd,
```
Can we just replace tex.rmsnorm_fwd with te_rmsnorm_fwd_triton? I thought we may need some careful fp8 output guarding?

wangye805 commented on transformer_engine/pytorch/ops/basic/rmsnorm.py:
```diff
@@ -205,8 +208,9 @@ def op_forward(
-        y, _, rstdevs = rmsnorm_fwd(
+        # Compute RMSNorm forward pass
+        rmsnorm_fwd_func = te_rmsnorm_fwd_triton if self.use_rmsnorm_triton else rmsnorm_fwd
+        y, _, rstdevs = rmsnorm_fwd_func(
```
Same here, could we running into fp8 output for this rmsnorm?

wangye805 commented on transformer_engine/pytorch/triton_kernels/rmsnorm.py:
```diff
@@ -297,3 +295,56 @@ def te_rmsnorm_bwd_triton(dz, x, rsigma, gamma, sm_margin, zero_centered_gamma):
+        )
+
+    #TODO: Update to include fp8 quantization.
+    if quantizer is not None:
```
Oh okay, you decided to error out when fp8 output is requested. In our previous versions, we fall back to hipified cuda kernels implementations for fp8 output instead of erroring out. Can we recover this behavior? Thanks

Micky774 committed:
Included HIP fallback

Micky774 committed:
Remove extra file

Micky774 commented:
@wangye805 I've updated the PR to include the HIP fallback. To test it, I've added a level 3 CI test run on `test_fusible_ops.py` with the Triton kernels enabled, since that test checks for correctness with FP8 quantization already. Furthermore, I added a temporary, explicit check on whether the appropriate warning is raised, which can be removed when we natively integrate FP8 quantization.

Micky774 committed:
Trigger Build

Micky774 committed:
Added type safety for ln_out

wangye805 commented on ci/pytorch.sh:
```diff
@@ -67,6 +67,7 @@ run_test_config(){
     run_default_fa 1 triton_kernels/test_norm_common.py
     #NVTE_USE_CAST_TRANSPOSE_TRITON=1 NVTE_USE_RMSNORM_TRITON=1 NVTE_USE_LAYERNORM_TRITON=1 run_default_fa 3 test_numerics.py
     NVTE_USE_RMSNORM_TRITON=1 NVTE_USE_LAYERNORM_TRITON=1 run_default_fa 3 test_numerics.py
+    NVTE_USE_RMSNORM_TRITON=1 run_default_fa 3 test_fusible_ops.py
```
This runtime warning capture can be set to level 1 CI
We also need a level 3 CI for test_numerics.py with this triton rmsnorm fwd. You can directly modify Line 69 above

Micky774 replied:
Done

wangye805 commented on tests/pytorch/test_fusible_ops.py:
```diff
@@ -1247,7 +1247,16 @@ def test_rmsnorm(
         )
         with te.fp8_autocast(enabled=quantized_compute, fp8_recipe=recipe):
-            y_test = forward(x_test)
+            # TODO: Remove when we support FP8 quantization in the rmsnorm
```
Please guard with IS_HIP_EXTENSION. We need to make sure that NV side not affected

Micky774 replied:
Done. I added a check for `NVTE_USE_RMSNORM_TRITON=1` as well, since otherwise we use the HIP kernel directly.

Micky774 committed:
Check for HIP, and adjust CI level

Micky774 committed:
Updated guard

Micky774 committed:
Merge branch 'dev' of https://github.com/ROCm/TransformerEngine into dev

Micky774 committed:
Used existing dtype map

Micky774 commented:
@wangye805 Ready for review
